drivers/firmware/ti_sci/ti_sci.c

```diff
@@ -0,0 +1,1660 @@
+	}
+
+	if (programmed_state) {
+		*programmed_state = resp.programmed_state;
+	}
+	if (current_state) {
+		*current_state = resp.current_state;
+	}
+	return ret;
+}
+
+int ti_sci_cmd_clk_is_auto(const struct device *dev, uint32_t dev_id, uint8_t clk_id,
+			   bool *req_state)
+{
+	uint8_t state = 0;
+	int ret;
+
+	if (!req_state) {
+		return -EINVAL;
+	}
+
+	if (!dev) {
+		return -EINVAL;
+	}
+
+	ret = ti_sci_cmd_get_clock_state(dev, dev_id, clk_id, &state, NULL);
+	if (ret) {
+		return ret;
+	}
+
+	*req_state = (state == MSG_CLOCK_SW_STATE_AUTO);
+	return 0;
+}
+
+int ti_sci_cmd_clk_is_on(const struct device *dev, uint32_t dev_id, uint8_t clk_id, bool *req_state,
+			 bool *curr_state)
+{
+	uint8_t c_state = 0, r_state = 0;
+	int ret;
+
+	if (!req_state && !curr_state) {
+		return -EINVAL;
+	}
+
+	if (!dev) {
+		return -EINVAL;
+	}
+
+	ret = ti_sci_cmd_get_clock_state(dev, dev_id, clk_id, &r_state, &c_state);
+	if (ret) {
+		return ret;
+	}
+
+	if (req_state) {
+		*req_state = (r_state == MSG_CLOCK_SW_STATE_REQ);
+	}
+	if (curr_state) {
+		*curr_state = (c_state == MSG_CLOCK_HW_STATE_READY);
+	}
+	return 0;
+}
+
+int ti_sci_cmd_clk_is_off(const struct device *dev, uint32_t dev_id, uint8_t clk_id,
+			  bool *req_state, bool *curr_state)
+{
+	uint8_t c_state = 0, r_state = 0;
+	int ret;
+
+	if (!req_state && !curr_state) {
+		return -EINVAL;
+	}
+
+	if (!dev) {
+		return -EINVAL;
+	}
+
+	ret = ti_sci_cmd_get_clock_state(dev, dev_id, clk_id, &r_state, &c_state);
+	if (ret) {
+		return ret;
+	}
+
+	if (req_state) {
+		*req_state = (r_state == MSG_CLOCK_SW_STATE_UNREQ);
+	}
+	if (curr_state) {
+		*curr_state = (c_state == MSG_CLOCK_HW_STATE_NOT_READY);
+	}
+	return 0;
+}
+
+int ti_sci_cmd_clk_get_match_freq(const struct device *dev, uint32_t dev_id, uint8_t clk_id,
+				  uint64_t min_freq, uint64_t target_freq, uint64_t max_freq,
+				  uint64_t *match_freq)
+{
+	struct ti_sci_msg_resp_query_clock_freq resp;
+	struct ti_sci_msg_req_query_clock_freq req;
+	struct ti_sci_xfer *xfer;
+	int ret = 0;
+
+	xfer = ti_sci_setup_one_xfer(dev, TI_SCI_MSG_QUERY_CLOCK_FREQ,
+				     TI_SCI_FLAG_REQ_ACK_ON_PROCESSED, (uint32_t *)&req,
+				     sizeof(req), &resp, sizeof(resp));
+
+	if (!xfer) {
+		LOG_ERR("Failed to setup transfer");
+		return -EINVAL;
+	}
+
+	req.dev_id = dev_id;
+	req.clk_id = clk_id;
+	req.min_freq_hz = min_freq;
+	req.target_freq_hz = target_freq;
+	req.max_freq_hz = max_freq;
+
+	ret = ti_sci_do_xfer(dev, xfer);
+	if (ret) {
+		return ret;
+	}
+
+	*match_freq = resp.freq_hz;
+
+	return 0;
+}
+
+int ti_sci_cmd_clk_set_freq(const struct device *dev, uint32_t dev_id, uint8_t clk_id,
+			    uint64_t min_freq, uint64_t target_freq, uint64_t max_freq)
+{
+	struct ti_sci_msg_req_set_clock_freq req;
+	struct ti_sci_xfer *xfer;
+	int ret = 0;
+
+	if (!dev) {
+		return -EINVAL;
+	}
+
+	xfer = ti_sci_setup_one_xfer(dev, TI_SCI_MSG_SET_CLOCK_FREQ,
+				     TI_SCI_FLAG_REQ_ACK_ON_PROCESSED, (uint32_t *)&req,
+				     sizeof(req), NULL, 0);
```

Comment:
You should always check if the xfer was setup correctly, otherwise you might send a NULL ptr to `ti_sci_do_xfer()`. Plus it is more consistent that way.